rfd/0162-machine-id-token-join-method-bot-instance.md

```diff
@@ -0,0 +1,653 @@
+this right when we can expect contention over a single bot instance resource.
+
+### CLI Changes
+
+#### `tbot`
+
+`tbot reset` will be added to allow a `tbot` instance to be reset. This will
+simply clear out any artifacts within the `tbot` storage directory.
+
+In addition, if the bot detects a change in join token or join method, it should
+automatically rotate it's keypair. This will ensure it presents as a fresh
+BotInstance to the AuthServer.
+
+A log message should be output that identifies the linked BotInstance at
+startup and on each heartbeat. This will allow users to easily correlate the
+`tbot` installation with a BotInstance.
+
+#### `tctl`
+
+Commands to list all BotInstances and the BotInstances for a specific Bot should
+be added to the `tctl bots` family:
+
+`tctl bots instances list`
+`tctl bots instances list --bot <bot name>`
+
+The `tctl tokens add` command should be extended to allow a new token to be
+associated with an existing Bot now that multiple Bot instances can be run
+against a single Bot:
+
+`tctl tokens add --type=bot --bot <bot name>`
+
+Additionally, `tctl rm`/`tctl get` should be able to operate on BotInstance.
+
+There is no requirement for it to be possible to create or update a BotInstance
+with `tctl`.
+
+### Analytics
+
+A PostHog event should be emitted for each BotInstance heartbeat. This will
+allow us to track active bots in a similar way to how we track active agents.
+
+```protobuf
+// a heartbeat for a Bot Instance
+//
+// PostHog event: tp.bot.instance.hb
+message BotInstanceHeartbeatEvent {
+  // anonymized name of the instance, 32 bytes (HMAC-SHA-256);
+  bytes bot_instance_name = 1;
+  // the version of tbot
+  string version = 2;
+  // indicates whether or not tbot is running in one-shot mode
+  bool one_shot = 3;
+  // indicates the configured join method of `tbot`.
+  string join_method = 4;
+}
+```
+
+Existing analytics for join, renewal and certificate generation should be
+extended to include the BotInstance ID anonymized. This will allow them to be
+linked together.
+
+### Migration/Compatability
+
+The "create if not exists" behaviour of the BotInstance resource will mean that
+existing Bot instances will have a BotInstance resource created on their first
+renewal after this feature is released. Their existing generation counter will
+be trusted on this first renewal. This allows for a seamless migration to the
+new system.
+
+Older `tbot` instances will not submit heartbeats. This means that their
+BotInstance will only contain authentication data. Any CLI or GUI that shows
+BotInstances should show a gracefully degraded state in this case that explains
+that the `tbot` needs to be upgraded.
+
+## Security Considerations
+
+### Audit Events
+
+An audit event should be added for the deletion of a BotInstance. The
+name of the BotInstance should be added to the existing join, renewal and
+certificate generation audit events.
```

Comment:
We should also put the bot instance's name/UUID in any existing audit events (like certificates issued) so that we have traceability into the actions of a particular bot.

In the future, we can then extend the access monitoring feature or even offer a bot details page in the UI that lists its renewals, join attempts, certs issued, etc.